src/EffectTransform.js

```diff
@@ -0,0 +1,102 @@
+/**
+ * @fileoverview
+ * A utility to transform a texture coordinate to another texture coordinate
+ * representing how the shaders apply effects.
+ */
+
+const twgl = require('twgl.js');
+
+const ShaderManager = require('./ShaderManager');
+
+/**
+ * A texture coordinate is between 0 and 1. 0.5 is the center position.
+ * @const {number}
+ */
+const CENTER_X = 0.5;
+
+/**
+ * A texture coordinate is between 0 and 1. 0.5 is the center position.
+ * @const {number}
+ */
+const CENTER_Y = 0.5;
+
+class EffectTransform {
+    /**
+     * Transform a texture coordinate to one that would be select after applying shader effects.
+     * @param {Drawable} drawable The drawable whose effects to emulate.
+     * @param {twgl.v3} vec The texture coordinate to transform.
+     * @param {?twgl.v3} dst A place to store the output coordinate.
+     * @return {twgl.v3} The coordinate after being transform by effects.
+     */
+    static transformPoint (drawable, vec, dst) {
```

Comment:
I'm not a huge fan of logically (though not quite literally) duplicating all this shader code here, but I guess it's worth it considering how much of a performance improvement we get from doing so. Still, I'd like to revisit this post-release.

Reply:
Same.